httpx.RequestNotRead: Attempted to access request content, without having called `read()`

I would like to test if a client uses the REST interface in a proper way. I wanted to use `respx` to mock the HTTP requests. I found no easy way to test the content of the HTTP request sent to the service.

What I would expect:
* I can use the `content` attribute on the `Request` object returned by the `calls` to verify correctness of the request.

What happens:
* Accessing the `content` attribute throws an `httpx.RequestNotRead` exception
* The workaround I have found is to access the `stream._body` attribute which is kind of a hack. Or maybe I don not use the library in a proper way and the documentation should show this kind of example.

Used Environment:
* Python 3.8.5
* pytest 6.0.2
* httpx 0.15.5
* respx 0.13.0

Here the minimal example code:
```python
import json
import respx
import httpx


def my_client():
    httpx.post("http://www.example.com/entries", content=json.dumps({"id": "1234567"}))
    return True


@respx.mock
def test_rest_call():
    """If the client sends the correct HTTP request - url and content."""
    respx.post(
        "http://www.example.com/entries",
        alias="example",
    )
    my_client()
    assert respx.aliases["example"].called
    assert respx.aliases["example"].call_count == 1
    request,_ = respx.aliases["example"].calls[0]
    
    # This works but is kinda ugly
    assert request.stream._body == '{"id": "1234567"}'.encode()
    # Would expect to work, but throws exception 'httpx.RequestNotRead'
    assert request.content == '{"id": "1234567"}'.encode()
```
Calling this with pytest:
```bash
$ pytest tests/test_request.py 
============================================================ test session starts =============================================================
platform linux -- Python 3.8.5, pytest-6.0.2, py-1.9.0, pluggy-0.13.1
rootdir: /home/max/work/iov42/python/core-sdk-python
plugins: xdoctest-0.15.0, typeguard-2.9.1
collected 1 item                                                                                                                             

tests/test_request.py F                                                                                                                [100%]

================================================================== FAILURES ==================================================================
_______________________________________________________________ test_rest_call _______________________________________________________________

    @respx.mock
    def test_rest_call():
        respx.post(
            "http://www.example.com/entries",
            alias="example",
        )
        my_client()
        assert respx.aliases["example"].called
        assert respx.aliases["example"].call_count == 1
        request,_ = respx.aliases["example"].calls[0]
    
        # This works but is kinda ugly
        assert request.stream._body == '{"id": "1234567"}'.encode()
        # Would expect to work, but does not work
>       assert request.content == '{"id": "1234567"}'.encode()

tests/test_request.py:26: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <Request('POST', 'http://www.example.com/entries')>

    @property
    def content(self) -> bytes:
        if not hasattr(self, "_content"):
>           raise RequestNotRead()
E           httpx.RequestNotRead: Attempted to access request content, without having called `read()`.

/home/max/.cache/pypoetry/virtualenvs/iov42-core-python-4oIM64sE-py3.8/lib/python3.8/site-packages/httpx/_models.py:823: RequestNotRead
========================================================== short test summary info ===========================================================
FAILED tests/test_request.py::test_rest_call - httpx.RequestNotRead: Attempted to access request content, without having called `read()`.
============================================================= 1 failed in 0.17s ==============================================================
```

--- a/respx/transports.py
+++ b/respx/transports.py
@@ -280,10 +280,11 @@ class BaseMockTransport:
         response = decode_response(response, request=request)
 
         # TODO: Skip recording stats for pass_through requests?
-        # Pre-read response, but only if mocked, not pass-through streams streams
+        # Pre-read request/response, but only if mocked, not for pass-through streams
         if response and not isinstance(
             response.stream, (SyncByteStream, AsyncByteStream)
         ):
+            request.read()
             response.read()
 
         if pattern:

--- a/tests/test_api.py
+++ b/tests/test_api.py
@@ -1,4 +1,5 @@
 import asyncio
+import json as jsonlib
 import os
 import re
 import socket
@@ -262,25 +263,33 @@ async def test_raising_content(client):
 async def test_callable_content(client):
     async with MockTransport() as respx_mock:
         url_pattern = re.compile(r"https://foo.bar/(?P<slug>\w+)/")
-        content = lambda request, slug: f"hello {slug}"
-        request = respx_mock.get(url_pattern, content=content)
 
-        async_response = await client.get("https://foo.bar/world/")
+        def content_callback(request, slug):
+            request.read()  # TODO: Make this not needed, might affect pass-through
+            content = jsonlib.loads(request.content)
+            return f"hello {slug}{content['x']}"
+
+        request = respx_mock.post(url_pattern, content=content_callback)
+
+        async_response = await client.post("https://foo.bar/world/", json={"x": "."})
         assert request.called is True
         assert async_response.status_code == 200
-        assert async_response.text == "hello world"
+        assert async_response.text == "hello world."
+        assert request.calls[-1][0].content == b'{"x": "."}'
 
         respx_mock.reset()
-        sync_response = httpx.get("https://foo.bar/world/")
+        sync_response = httpx.post("https://foo.bar/jonas/", json={"x": "!"})
         assert request.called is True
         assert sync_response.status_code == 200
-        assert sync_response.text == "hello world"
+        assert sync_response.text == "hello jonas!"
+        assert request.calls[-1][0].content == b'{"x": "!"}'
 
 
 @pytest.mark.asyncio
 async def test_request_callback(client):
     def callback(request, response):
-        if request.url.host == "foo.bar":
+        request.read()
+        if request.url.host == "foo.bar" and request.content == b'{"foo": "bar"}':
             response.headers["X-Foo"] = "bar"
             response.content = lambda request, name: f"hello {name}"
             response.context["name"] = "lundberg"
@@ -289,7 +298,7 @@ async def test_request_callback(client):
 
     async with MockTransport(assert_all_called=False) as respx_mock:
         request = respx_mock.add(callback, status_code=202, headers={"X-Ham": "spam"})
-        response = await client.get("https://foo.bar/")
+        response = await client.post("https://foo.bar/", json={"foo": "bar"})
 
         assert request.called is True
         assert request.pass_through is None
@@ -354,17 +363,29 @@ async def test_pass_through(client, parameters, expected):
 @pytest.mark.asyncio
 async def test_external_pass_through(client):  # pragma: nocover
     with respx.mock:
-        url = "https://example.org/"
+        # Mock pass-through call
+        url = "https://httpbin.org/post"
         respx.get(url, content=b"", pass_through=True)
-        response = await client.get(url)
+
+        # Mock a non-matching callback pattern pre-reading request data
+        def callback(req, res):
+            req.read()  # TODO: Make this not needed, might affect pass-through
+            assert req.content == b'{"foo": "bar"}'
+            return None
+
+        respx.add(callback)
+
+        # Make external pass-through call
+        response = await client.post(url, json={"foo": "bar"})
 
         assert response.content is not None
         assert len(response.content) > 0
         assert "Content-Length" in response.headers
         assert int(response.headers["Content-Length"]) > 0
+        assert response.json()["json"] == {"foo": "bar"}
 
         _, resp = respx.calls[-1]
-        await resp.aread()
+        await resp.aread()  # Read async pass-through response
         assert resp.content == b"", "Should be 0, stream already read by real Response!"
         assert "Content-Length" in resp.headers
         assert int(resp.headers["Content-Length"]) > 0

